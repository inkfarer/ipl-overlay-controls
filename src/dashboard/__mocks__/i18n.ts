import i18next from 'i18next';
import { config } from '@vue/test-utils';

jest.mock('i18next-vue', () => ({
    __esModule: true,
    useTranslation: () => ({ t: i18next.t })
}));

beforeAll(async () => {
    await i18next.init({
        lng: 'cimode',
        appendNamespaceToCIMode: true
    });
    config.global.mocks = {
        $t: i18next.t
    };
});
