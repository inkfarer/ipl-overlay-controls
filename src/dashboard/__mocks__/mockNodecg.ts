import Mock = jest.Mock;

export type ReplicantChangeHandler = (newValue?: unknown, oldValue?: unknown) => void;
export type MockReplicant = { value: unknown };
type OnFunction = (event: string, handler: ReplicantChangeHandler) => void;

export class MockNodecg {
    replicants: {[key: string]: MockReplicant};
    listeners: {[key: string]: ReplicantChangeHandler};
    sendMessage: Mock

    constructor() {
        this.replicants = {};
        this.listeners = {};
        this.sendMessage = jest.fn();
    }

    init(): void {
        const self = this;
        window.nodecg = {
            // @ts-ignore: Just a mock
            Replicant(name: string) {
                const replicantValue: { value: unknown, on: OnFunction } = {
                    value: undefined,
                    on: (event: string, handler: ReplicantChangeHandler) => {
                        self.listeners[name] = handler;
                    }
                };
                self.replicants[name] = replicantValue;
                return replicantValue;
            },
            sendMessage: self.sendMessage,
            // @ts-ignore: Just a mock
            log: {
                info: jest.fn()
            }
        };

        // @ts-ignore: It's a mock.
        window.NodeCG = {
            waitForReplicants: jest.fn().mockResolvedValue(true)
        };
    }
}
