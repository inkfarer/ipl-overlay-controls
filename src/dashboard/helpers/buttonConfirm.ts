/*
buttonConfirm

After importing this script to a dashboard panel, add the attribute 'data-requires-confirmation' to any button.
The button will now ask for confirmation when it is first clicked.
When it is clicked a second time, a 'confirm' event is sent by it, which can be listened for by scripts.

Example:

HTML:
    <button id="dangerous-action-button" data-requires-confirmation>DO A THING</button>
JS:
    document.getElementById('dangerous-action-button').addEventListener('confirm', () => {
        // do something
    });
*/

window.addEventListener('DOMContentLoaded', () => {
    function addListener(btn: HTMLButtonElement) {
        if (btn.dataset.hasConfirmListener !== 'true') {
            btn.addEventListener('click', e => {
                const target = e.target as HTMLButtonElement;

                function resetButton() {
                    target.dataset.isClicked = 'false';
                    target.innerText = target.dataset.originalText;
                    target.style.backgroundColor = '';
                }

                if (target.dataset.isClicked === 'true') {
                    target.dispatchEvent(new Event('confirm'));
                    resetButton();
                } else {
                    target.dataset.isClicked = 'true';
                    target.dataset.originalText = target.innerText;
                    target.style.backgroundColor = 'var(--red)';
                    target.innerText = 'are you sure?';

                    setTimeout(() => {
                        resetButton();
                    }, 5000);
                }
            });

            btn.dataset.hasConfirmListener = 'true';
        }
    }

    const buttons = document.querySelectorAll('button[data-requires-confirmation]');
    buttons.forEach(addListener);

    const observer = new MutationObserver(mutations => {
        mutations.forEach(mutation => {
            mutation.addedNodes.forEach(node => {
                if (node.nodeType === Node.ELEMENT_NODE
                    && node.nodeName === 'BUTTON'
                    && (node as HTMLElement).dataset.requiresConfirmation) {

                    const btn = node as HTMLButtonElement;
                    addListener(btn);
                }
            });
        });
    });

    observer.observe(document.body, {
        subtree: true,
        childList: true
    });
});
