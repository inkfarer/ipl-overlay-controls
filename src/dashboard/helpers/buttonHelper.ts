import { colors } from '../styles/colors';

export function setToggleButtonDisabled(
    toggleTrue: HTMLButtonElement,
    toggleFalse: HTMLButtonElement,
    state: boolean
): void {
    toggleTrue.disabled = state;
    toggleFalse.disabled = !state;
}

export function addChangeReminder(
    elements: HTMLElement[] | NodeListOf<HTMLElement>,
    updateButton: HTMLButtonElement
): void {
    elements.forEach(element => {
        if (!element.tagName) return;

        let event;
        if (element.tagName.toLowerCase() === 'input') event = 'input';
        else if (element.tagName.toLowerCase() === 'select') event = 'change';
        else return;

        element.addEventListener(event, e => {
            const target = e.target as HTMLElement;
            target.dataset.edited = 'true';
            updateButton.style.backgroundColor = colors.red;
        });
    });

    updateButton.addEventListener('click', () => {
        elements.forEach(elem => {
            elem.dataset.edited = 'false';
        });
        updateButton.style.backgroundColor = colors.blue;
    });
}
