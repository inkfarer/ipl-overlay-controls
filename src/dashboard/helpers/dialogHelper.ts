import { NodecgDialog } from '../types/dialog';

export function closeDialog(dialogName: string): void {
    (nodecg.getDialog(dialogName) as NodecgDialog).close();
}
