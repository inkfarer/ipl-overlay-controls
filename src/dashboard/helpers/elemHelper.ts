export function appendChildren(element: HTMLElement, ...children: HTMLElement[]): void {
    children.forEach(elem => {
        element.appendChild(elem);
    });
}

export function elementById<T extends HTMLElement>(id: string): T | undefined {
    return document.getElementById(id) as T;
}

export function dispatch(element: EventTarget, event: string): void {
    element.dispatchEvent(new Event(event));
}

export function hideElement(element: HTMLElement): void {
    element.style.display = 'none';
}

export function showElement(element: HTMLElement): void {
    element.style.removeProperty('display');
}

export function prependToBody(elem: HTMLElement): void {
    document.body.prepend(elem);
}

export function appendElementAfter(newElement: HTMLElement, element: HTMLElement): void {
    element.parentNode.insertBefore(newElement, element.nextSibling);
}
