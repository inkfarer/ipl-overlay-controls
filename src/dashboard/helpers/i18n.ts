import { RuntimeConfig } from 'schemas';
import i18next, { type BackendModule } from 'i18next';

const I18nBackend: BackendModule = {
    type: 'backend',

    // eslint-disable-next-line @typescript-eslint/no-empty-function
    init() {

    },

    async read(language, namespace) {
        return loadTranslation(language, namespace);
    },
};

export async function initI18n(namespaces: string | string[]): Promise<void> {
    const ns = Array.isArray(namespaces) ? [...namespaces, 'common']: [namespaces, 'common'];
    await i18next
        .use(I18nBackend)
        .init({
            lng: 'en',
            fallbackLng: 'en',
            ns,
            interpolation: {
                escapeValue: false
            }
        });

    const runtimeConfig = nodecg.Replicant<RuntimeConfig>('runtimeConfig');
    runtimeConfig.on('change', async (newValue, oldValue) => {
        if (!oldValue || newValue.interfaceLocale !== oldValue.interfaceLocale) {
            await i18next.changeLanguage(newValue.interfaceLocale.toLowerCase()).catch(e => {
                console.error('Failed to change interface language', e);
            });
            if (i18next.exists('title')) {
                try {
                    window.frameElement.parentElement.setAttribute('display-title', i18next.t('title'));
                } catch (ignore) {

                }
            }
        }
    });
}

async function loadTranslation(locale: string, name: string) {
    try {
        return await import(`../../helpers/i18n/${locale.toLowerCase()}/${name}.json`);
    } catch (ignore) {
        return null;
    }
}
