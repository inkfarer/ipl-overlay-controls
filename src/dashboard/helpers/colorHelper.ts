import { colors, getColorOptionName } from '../../helpers/splatoonData';

export function fillColorSelector(select: HTMLSelectElement): void {
    for (let i = 0; i < colors.length; i++) {
        const element = colors[i];

        const optGroup = document.createElement('optgroup');
        optGroup.label = element.meta.name;

        for (let j = 0; j < element.colors.length; j++) {
            const color = element.colors[j];

            const option = document.createElement('option');
            option.value = getColorOptionName(color.index, element.meta.name);
            option.text = color.title;
            option.dataset.index = String(color.index);
            option.dataset.firstColor = color.clrA;
            option.dataset.secondColor = color.clrB;
            option.dataset.categoryName = element.meta.name;
            option.disabled = color.isCustom;

            optGroup.appendChild(option);
        }

        select.appendChild(optGroup);
    }
}

export function getContrastingTextColor(backgroundColor: string): string {
    const { r, g, b } = hexToRbg(backgroundColor);
    const yiq = ((r * 299) + (g * 587) + (b * 114)) / 1000;
    return (yiq >= 128) ? '#333' : 'white';
}

export function hexToRbg(color: string): { r: number, g: number, b: number } {
    color = color.replace('#', '');
    return {
        r: parseInt(color.substr(0, 2), 16),
        g: parseInt(color.substr(2, 2), 16),
        b: parseInt(color.substr(4, 2), 16)
    };
}
