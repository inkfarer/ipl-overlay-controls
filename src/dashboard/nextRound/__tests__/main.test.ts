import { MockNodecg } from '../../__mocks__/mockNodecg';
import { dispatch, elementById } from '../../helpers/elemHelper';
import { Module } from '../../../helpers/__mocks__/module';

describe('main', () => {
    const mockTeamImageToggleHelper = {
        handleTeamImageToggleChange: jest.fn()
    };
    let main: Module;
    let nodecg: MockNodecg;

    jest.mock('../../helpers/teamImageToggleHelper', () => mockTeamImageToggleHelper);

    beforeEach(() => {
        jest.resetModules();
        jest.resetAllMocks();

        nodecg = new MockNodecg();
        nodecg.init();

        document.body.innerHTML = `
            <button id="update-next-round-btn"></button>
            <button id="begin-next-match-btn"></button>
            <select id="round-selector"></select>
            <select id="next-team-a-selector" class="team-selector"></select>
            <select id="next-team-b-selector" class="team-selector"></select>
            <div id="saved-progress-message">
                <div class="content"></div>
            </div>
            <input id="show-team-a-image">
            <input id="show-team-b-image">
        `;

        main = require('../main');
    });

    describe('roundStore: change', () => {
        beforeEach(() => {
            nodecg.replicants.nextRound.value = { round: { id: 'bbbbbb' } };
        });

        it('updates selector', () => {
            nodecg.listeners.roundStore({
                aaaaaa: {
                    meta: { name: 'Round 1' }
                },
                bbbbbb: {
                    meta: { name: 'Round 2' }
                }
            });

            expect(elementById('round-selector').innerHTML).toMatchSnapshot();
        });
    });

    describe('nextRound: change', () => {
        beforeEach(() => {
            nodecg.replicants.roundStore.value = {};
        });

        it('updates inputs', () => {
            const teamASelector = elementById<HTMLSelectElement>('next-team-a-selector');
            const teamBSelector = elementById<HTMLSelectElement>('next-team-b-selector');
            teamASelector.innerHTML = '<option value="asdasd"></option><option value="123123"></option>';
            teamBSelector.innerHTML = '<option value="qweqwe"></option><option value="456456"></option>';
            const roundSelector = elementById<HTMLSelectElement>('round-selector');
            roundSelector.innerHTML = '<option value="zxczxc"></option><option value="890890"></option>';

            nodecg.listeners.nextRound({
                teamA: { showLogo: true, id: '123123' },
                teamB: { showLogo: false, id: '456456' },
                round: { id: '890890' }
            });

            const showTeamAImage = elementById<HTMLInputElement>('show-team-a-image');
            const showTeamBImage = elementById<HTMLInputElement>('show-team-b-image');
            expect(showTeamAImage.checked).toEqual(true);
            expect(showTeamAImage.dataset.teamId).toEqual('123123');
            expect(showTeamBImage.checked).toEqual(false);
            expect(showTeamBImage.dataset.teamId).toEqual('456456');
            expect(teamASelector.value).toEqual('123123');
            expect(teamBSelector.value).toEqual('456456');
            expect(roundSelector.value).toEqual('890890');
        });
    });

    describe('tournamentData: change', () => {
        beforeEach(() => {
            nodecg.replicants.nextRound.value = {
                teamA: { id: '111' },
                teamB: { id: '222' }
            };
        });

        it('updates inputs', () => {
            const teamASelector = elementById('next-team-a-selector');
            const teamBSelector = elementById('next-team-b-selector');
            teamASelector.innerHTML = '<option value="valuevalue"></option>';
            teamBSelector.innerHTML = '<option value="value2value2"></option>';

            nodecg.listeners.tournamentData({
                teams: [
                    { name: 'Team One', id: '111' },
                    { name: 'Team Two', id: '222' }
                ]
            });

            expect(teamASelector.innerHTML).toMatchSnapshot();
            expect(teamBSelector.innerHTML).toMatchSnapshot();
        });
    });

    describe('update-next-round-btn: click', () => {
        it('sends a message', () => {
            const teamASelector = elementById<HTMLSelectElement>('next-team-a-selector');
            const teamBSelector = elementById<HTMLSelectElement>('next-team-b-selector');
            teamASelector.innerHTML = '<option value="aaa"></option>';
            teamASelector.value = 'aaa';
            teamBSelector.innerHTML = '<option value="bbb"></option>';
            teamBSelector.value = 'bbb';
            const roundSelector = elementById<HTMLSelectElement>('round-selector');
            roundSelector.innerHTML = '<option value="roundround"></option>';
            roundSelector.value = 'roundround';

            dispatch(elementById('update-next-round-btn'), 'click');

            expect(nodecg.sendMessage).toHaveBeenCalledWith('setNextRound', {
                teamAId: 'aaa',
                teamBId: 'bbb',
                roundId: 'roundround'
            });
        });
    });

    describe('begin-next-match-btn: confirm', () => {
        it('sends a message', () => {
            dispatch(elementById('begin-next-match-btn'), 'confirm');

            expect(nodecg.sendMessage).toHaveBeenCalledWith('beginNextMatch');
        });
    });

    describe('show-team-a-image: change', () => {
        it('handles change', () => {
            dispatch(elementById('show-team-a-image'), 'change');

            expect(mockTeamImageToggleHelper.handleTeamImageToggleChange).toHaveBeenCalled();
        });
    });

    describe('show-team-b-image: change', () => {
        it('handles change', () => {
            dispatch(elementById('show-team-b-image'), 'change');

            expect(mockTeamImageToggleHelper.handleTeamImageToggleChange).toHaveBeenCalled();
        });
    });

    describe('checkNextRoundProgress', () => {
        it('shows message if round has progress', () => {
            const message = elementById('saved-progress-message');
            const messageContent = message.querySelector('.content') as HTMLElement;

            main.checkNextRoundProgress({
                meta: { name: 'Round Round' },
                teamA: { score: 1, name: 'Team Alpha' },
                teamB: { score: 0, name: 'Team Bravo' }
            });

            expect(message.style.display).toEqual('');
            expect(messageContent.innerText).toEqual('\'Round Round\' already has saved progress.\n'
                + '(Team Alpha vs Team Bravo)');
        });

        it('shows message if round is completed', () => {
            const message = elementById('saved-progress-message');
            const messageContent = message.querySelector('.content') as HTMLElement;

            main.checkNextRoundProgress({
                meta: { name: 'Round Round 2', isCompleted: true },
                teamA: { score: 1, name: 'Team Gamma' },
                teamB: { score: 2, name: 'Team Kappa' }
            });

            expect(message.style.display).toEqual('');
            expect(messageContent.innerText).toEqual('\'Round Round 2\' is already completed.\n'
                + '(Team Gamma vs Team Kappa)');
        });

        it('hides message if round has no progress', () => {
            const message = elementById('saved-progress-message');

            main.checkNextRoundProgress({
                meta: { name: 'Round Round 2', isCompleted: false },
                teamA: { score: 0, name: 'Team Gamma' },
                teamB: { score: 0, name: 'Team Kappa' }
            });

            expect(message.style.display).toEqual('none');
        });
    });
});
