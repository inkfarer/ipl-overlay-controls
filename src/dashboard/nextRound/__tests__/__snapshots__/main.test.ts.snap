// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`main roundStore: change updates selector 1`] = `"<option value=\\"aaaaaa\\">Round 1</option><option value=\\"bbbbbb\\">Round 2</option>"`;

exports[`main tournamentData: change updates inputs 1`] = `"<option value=\\"111\\">Team One</option><option value=\\"222\\">Team Two</option>"`;

exports[`main tournamentData: change updates inputs 2`] = `"<option value=\\"111\\">Team One</option><option value=\\"222\\">Team Two</option>"`;
